src/fb.c

```diff
@@ -1,63 +1,275 @@
 /* SPDX-License-Identifier: MIT */
 
+#include <assert.h>
+
 #include "fb.h"
+#include "smp.h"
+#include "string.h"
 #include "utils.h"
 #include "xnuboot.h"
 
-u32 *fb;
-int fb_s, fb_w, fb_h, fb_d;
-int scale = 1;
+#define FB_DEPTH_FLAG_RETINA 0x10000
+#define FB_DEPTH_MASK        0xff
+
+#define CONSOLE_MAX_ROWS 60
+#define CONSOLE_MAX_COLS 240
+
+#define CONSOLE_MARGIN_LEFT   30
+#define CONSOLE_MARGIN_RIGHT  30
+#define CONSOLE_MARGIN_TOP    50
+#define CONSOLE_MARGIN_BOTTOM 50
+
+fb_t fb;
+
+typedef struct {
+    u32 row;
+    u32 col;
+} console_pos_t;
+
+static struct {
+    u32 *ptr;
+    u32 width;
+    u32 height;
+    u32 x;
+    u32 y;
+
+    u32 enabled;
+} logo = {.enabled = 0};
+
+static struct {
+    u32 initialized;
+
+    struct {
+        u8 *ptr;
+        u32 width;
+        u32 height;
+    } font;
 
-static u32 *logo;
-static int logo_w, logo_h;
+    console_pos_t cursor;
+
+    u32 n_rows;
+    u32 n_cols;
+    u32 row_offset;
+
+    u8 text[CONSOLE_MAX_ROWS][CONSOLE_MAX_COLS];
+} console = {.initialized = 0,
+             .cursor = {.row = 0, .col = 0},
+             .n_rows = CONSOLE_MAX_ROWS,
+             .n_cols = CONSOLE_MAX_COLS,
+             .row_offset = 0,
+             .text = {{0}}};
 
 extern char _binary_build_bootlogo_128_bin_start[0];
 extern char _binary_build_bootlogo_256_bin_start[0];
```

Comment:
While you're working on this you might as well change these to match (using `u8` and `[]` instead of `char` and `[0]`)

Reply:
good point, will change these as well.